File: src/lib/coach/prompt.ts
import { readFileSync } from "node:fs";
import path from "node:path";

/**
 * The coach persona lives in shared/coach/system-prompt.md so the Next.js
 * runtime and the FastAPI/LangGraph service read the *same bytes* — a forked
 * copy would let the two backends answer differently within a week.
 *
 * The file's editorial header (everything above the first `---` rule) is for
 * humans; only the prose below it is sent to the model.
 */
const PROMPT_PATH = path.join(process.cwd(), "shared", "coach", "system-prompt.md");

let cached: string | null = null;

export function coachSystemPrompt(): string {
  if (cached) return cached;

  const raw = readFileSync(PROMPT_PATH, "utf8");
  const marker = "\n---\n";
  const idx = raw.indexOf(marker);
  cached = (idx === -1 ? raw : raw.slice(idx + marker.length)).trim();
  return cached;
}

File: backend/coach/prompt.py
"""Loads the coach persona from the repo-shared markdown file.

Deliberately the same file the Next.js backend reads
(``shared/coach/system-prompt.md``) — two copies of a system prompt drift
within a week and the two backends start answering differently.

The file's editorial header (everything above the first ``---`` rule) is for
humans; only the prose below it is sent to the model.
"""

from __future__ import annotations

import os
from functools import lru_cache
from pathlib import Path

# backend/coach/prompt.py -> backend/coach -> backend -> repo root
REPO_ROOT = Path(__file__).resolve().parents[2]
DEFAULT_PROMPT_PATH = REPO_ROOT / "shared" / "coach" / "system-prompt.md"


@lru_cache(maxsize=1)
def coach_system_prompt() -> str:
    path = Path(os.getenv("COACH_PROMPT_PATH", DEFAULT_PROMPT_PATH))
    raw = path.read_text(encoding="utf-8")
    marker = "\n---\n"
    _, sep, body = raw.partition(marker)
    return (body if sep else raw).strip()
